### ISSUE TITLE
make hinoki figure out lifetimes automatically by looking at the dependencies

### ISSUE THREAD

**Issue by snd** [snd/hinoki#4]
currently the user has to manually divide/organize factories by their lifetime (the folders `application`, `request`, `middleware` in the case of [fragments](https://github.com/snd/fragments)) and move factories manually between lifetimes/files if their lifetime changes.

it would be really nice if hinoki would figure this out for us !

that is possible:
if a fragment directly or indirectly depends on `params` or `next` it is in the **middleware** lifetime.
if a fragment directly or indirectly depends on `req` or `res` it is in the **request** lifetime.
otherwise it is in the **application** lifetime.

is this a good idea overall ?
are there any disadvantages/showstoppers when compared to the manual solution ?
does this break other usage patterns of hinoki than our own ?
how to implement this ?
how to transition to this ?

**needs a lot more thought !**

### possible implementation

factories/sources are currently complected with state/values in containers.

decomplect: have hinoki consume an array of sources (which are either "objects containing factories" or "factory factories") as well as an array of lifetimes.

lifetimes will be what `lifetime.values` currently is.
lifetimes are simply objects of values. 
lifetimes manage state.

**a value will then be cached in the last lifetime that contains values it directly or indirectly depends on. et voila !**

``` javascript
hinoki(sourceS, lifetimeS, nameS);
```

### notes

care has to be taken with fragments that should only exist during a specific lifetime but don't depend on values in that lifetime.
`k` in fragments is an example.
a solution is to use `params` or `next` as an unused "marker-dependency" to force it into the **middleware** lifetime.

need a solution for `promisesAwaitingResolution`.
previously in lifetimes. where now ?
as `__promises` property in the new lifetimes ?
`__promises` will then be a protected property that is not an allowed **name**.

we could also detect lifetimes ahead of time by looking at the dependencies
and then either add them as metadata to the factories `$lifetime` or filter factories into the specific lifetimes and use the old lifetime system.

lifetimes only control/influence where and how long the value will be cached.

another advantage is that bundling fragments in node modules becomes much easier this way.

### how will hinoki figure out the lifetimes?

illustrated on the following example:

``` javascript
var fragmentsPostgres = require('fragments-postgres');
var hinoki = require('hinoki');

var factories = {
  canAccessUsers: function(currentUserHasRight) { /* ... */ },
  currentUserHasRight: function(currentRightsObject) { /* ... */ },
  currentRightsObject: function(currentUser) { /* ... */ },
  currentUser: function(firstUserWhereId, token) { /* ... */ },
  token: function(req, res) { /* ... */ },
  mesa: function(databaseUrl) { /* ... */ },
  databaseUrl: function(env) { /* ... */ },
};

var applicationLifetime = {};
var requestLifetime = {
  req: /* ... */,
  res: /* ... */,
};

hinoki(
  [factories, fragmentsPostgres], 
  [applicationLifetime, requestLifetime], 
  'canAccessUsers',
).then(function(canAccessUsers) { /* ... * / });
```

now, what happens on the last line ?

hinoki tries to look up `canAccessUsers` as a value by looking into each lifetime in order:
if a value is found then it is returned. in this case no value is found.

next hinoki calls the `source` function with `'canAccessUsers'`:
the merged source calls each of the merged source functions with `'canAccessUsers'` in order.
the source `hinoki.source.fromObject(factories)` returns a factory.

for each factory whose value should be cached (no `$nocache` property set) the index of the lifetime
that is the cache target (`cacheTarget`) will be maintained during the next steps. `cacheTarget` is initialized to `0`, meaning "the first lifetime".

hinoki figures out the dependencies of the factory (by looking at the `$inject` property then by parsing from the stringified factory)

the `cacheTarget` of a factory is the highest `cacheTarget` of its dependencies.
the `cacheTarget` of a value is the index of the lifetime that owns the value.
the `cacheTarget` of a factory with zero dependencies is `0`.

*`cacheTarget` indexes are only valid during a single injection cycle (while the lifetimes stay the same)*

the max (highest number) `cacheTarget` is chosen.
for pending promises the `cacheTarget` is remembered in the closure of the promise.

`cacheTarget` is passed along with each injection pass.

once the value is resolved: `lifetimes[cacheTarget] = value;`

this should have zero runtime overhead: it just requires a few numbers being passed around and compared during injection.

for the above example `cacheTarget` is initally `0` and becomes `1` for all those values that directly and indirectly depend on `req` or `res`.

**Comment by sh** [comment]
:+1: 

> care has to be taken with fragments that should only exist during a specific lifetime but don't depend on values in that lifetime.
> k in fragments is an example.
> a solution is to use params or next as an unused "marker-dependency" to force it into the middleware lifetime.

fix via metadata e.g. ````module.exports.someFoo.$lifetime = "middleware"```` / ````module.exports.someFoo.$lifetimeIsMiddleware = true````

**Comment by snd** [comment]
:+1:

**Comment by snd** [comment]
this is now present in master and released > 1.0.0-rc.1
